File: agents/tools/config_diff.py
"""config diff computation.

compares two config blobs field by field and produces atomic change
dicts suitable for insertion into the config_changelog table.
"""

from __future__ import annotations

from typing import Any


# valid change categories matching the config_changelog.change_category enum
CHANGE_CATEGORIES = {
    "knob_tuned",
    "tool_enabled",
    "tool_disabled",
    "weight_adjusted",
    "threshold_adjusted",
    "session_rule_changed",
    "scoring_changed",
}


def compute_config_diff(
    old_config: dict[str, Any],
    new_config: dict[str, Any],
) -> list[dict[str, Any]]:
    """compare two config blobs and produce a list of atomic changes.

    each change dict has keys: change_category, target_timescale,
    target_tool_id, target_tool_type, target_param, old_value,
    new_value, reason.
    """
    changes: list[dict[str, Any]] = []

    # compare tickers
    old_tickers = set(old_config.get("tickers", []))
    new_tickers = set(new_config.get("tickers", []))
    if old_tickers != new_tickers:
        changes.append(_change(
            category="session_rule_changed",
            param="tickers",
            old_value=sorted(old_tickers),
            new_value=sorted(new_tickers),
        ))

    # compare scoring
    changes.extend(_diff_scoring(
        old_config.get("scoring", {}),
        new_config.get("scoring", {}),
    ))

    # compare session rules
    changes.extend(_diff_session(
        old_config.get("session", {}),
        new_config.get("session", {}),
    ))

    # compare indicators
    changes.extend(_diff_tools(
        old_config.get("indicators", []),
        new_config.get("indicators", []),
        tool_kind="indicator",
    ))

    # compare actions
    changes.extend(_diff_tools(
        old_config.get("actions", []),
        new_config.get("actions", []),
        tool_kind="action",
    ))

    return changes


def _change(
    category: str,
    param: str,
    old_value: Any = None,
    new_value: Any = None,
    timescale: str | None = None,
    tool_id: str | None = None,
    tool_type: str | None = None,
    reason: str = "config mutation by PM agent",
) -> dict[str, Any]:
    return {
        "change_category": category,
        "target_timescale": timescale,
        "target_tool_id": tool_id,
        "target_tool_type": tool_type,
        "target_param": param,
        "old_value": old_value,
        "new_value": new_value,
        "reason": reason,
    }


def _diff_scoring(old: dict, new: dict) -> list[dict[str, Any]]:
    changes: list[dict[str, Any]] = []

    # thresholds
    for key in ("entry_threshold", "exit_threshold"):
        old_val = old.get(key)
        new_val = new.get(key)
        if old_val != new_val:
            changes.append(_change(
                category="threshold_adjusted",
                param=key,
                old_value=old_val,
                new_value=new_val,
            ))

    # aggregation method
    if old.get("aggregation") != new.get("aggregation"):
        changes.append(_change(
            category="scoring_changed",
            param="aggregation",
            old_value=old.get("aggregation"),
            new_value=new.get("aggregation"),
        ))

    # hard gate timescales
    old_gates = old.get("hard_gate_timescales", [])
    new_gates = new.get("hard_gate_timescales", [])
    if sorted(str(g) for g in old_gates) != sorted(str(g) for g in new_gates):
        changes.append(_change(
            category="scoring_changed",
            param="hard_gate_timescales",
            old_value=old_gates,
            new_value=new_gates,
        ))

    # timescale weights
    old_weights = old.get("timescale_weights", {})
    new_weights = new.get("timescale_weights", {})
    all_timescales = set(list(old_weights.keys()) + list(new_weights.keys()))
    for ts in sorted(all_timescales):
        old_w = old_weights.get(ts)
        new_w = new_weights.get(ts)
        if old_w != new_w:
            changes.append(_change(
                category="weight_adjusted",
                param="timescale_weight",
                timescale=ts,
                old_value=old_w,
                new_value=new_w,
            ))

    return changes


def _diff_session(old: dict, new: dict) -> list[dict[str, Any]]:
    changes: list[dict[str, Any]] = []
    all_keys = set(list(old.keys()) + list(new.keys()))
    for key in sorted(all_keys):
        if old.get(key) != new.get(key):
            changes.append(_change(
                category="session_rule_changed",
                param=key,
                old_value=old.get(key),
                new_value=new.get(key),
            ))
    return changes


def _diff_tools(
    old_tools: list[dict],
    new_tools: list[dict],
    tool_kind: str,
) -> list[dict[str, Any]]:
    """diff indicator or action tool lists by instance_id."""
    changes: list[dict[str, Any]] = []

    old_by_id = {t["instance_id"]: t for t in old_tools if "instance_id" in t}
    new_by_id = {t["instance_id"]: t for t in new_tools if "instance_id" in t}

    all_ids = set(list(old_by_id.keys()) + list(new_by_id.keys()))

    for tool_id in sorted(all_ids):
        old_tool = old_by_id.get(tool_id)
        new_tool = new_by_id.get(tool_id)

        if old_tool is None and new_tool is not None:
            # new tool added (enabled)
            changes.append(_change(
                category="tool_enabled",
                param="enabled",
                tool_id=tool_id,
                tool_type=new_tool.get(f"{tool_kind}_type", new_tool.get("action_type", new_tool.get("indicator_type"))),
                timescale=new_tool.get("timescale"),
                old_value=None,
                new_value=True,
            ))
            continue

        if old_tool is not None and new_tool is None:
            # tool removed (disabled)
            changes.append(_change(
                category="tool_disabled",
                param="enabled",
                tool_id=tool_id,
                tool_type=old_tool.get(f"{tool_kind}_type", old_tool.get("action_type", old_tool.get("indicator_type"))),
                timescale=old_tool.get("timescale"),
                old_value=True,
                new_value=None,
            ))
            continue

        # both exist — check for changes
        assert old_tool is not None and new_tool is not None

        tool_type = new_tool.get(f"{tool_kind}_type", new_tool.get("action_type", new_tool.get("indicator_type")))
        timescale = new_tool.get("timescale")

        # enabled/disabled toggle
        old_enabled = old_tool.get("enabled", True)
        new_enabled = new_tool.get("enabled", True)
        if old_enabled != new_enabled:
            cat = "tool_enabled" if new_enabled else "tool_disabled"
            changes.append(_change(
                category=cat,
                param="enabled",
                tool_id=tool_id,
                tool_type=tool_type,
                timescale=timescale,
                old_value=old_enabled,
                new_value=new_enabled,
            ))

        # weight change (indicators only)
        if "weight" in old_tool or "weight" in new_tool:
            old_w = old_tool.get("weight")
            new_w = new_tool.get("weight")
            if old_w != new_w:
                changes.append(_change(
                    category="weight_adjusted",
                    param="weight",
                    tool_id=tool_id,
                    tool_type=tool_type,
                    timescale=timescale,
                    old_value=old_w,
                    new_value=new_w,
                ))

        # params diff
        old_params = old_tool.get("params", {})
        new_params = new_tool.get("params", {})
        all_param_keys = set(list(old_params.keys()) + list(new_params.keys()))
        for pkey in sorted(all_param_keys):
            if old_params.get(pkey) != new_params.get(pkey):
                changes.append(_change(
                    category="knob_tuned",
                    param=pkey,
                    tool_id=tool_id,
                    tool_type=tool_type,
                    timescale=timescale,
                    old_value=old_params.get(pkey),
                    new_value=new_params.get(pkey),
                ))

    return changes

File: agents-ts/src/tools/config-diff.ts
/**
 * config diff computation.
 *
 * compares two config blobs field by field and produces atomic change
 * dicts suitable for insertion into the config_changelog table.
 */

// valid change categories matching the config_changelog.change_category enum
export const CHANGE_CATEGORIES = new Set([
  "knob_tuned",
  "tool_enabled",
  "tool_disabled",
  "weight_adjusted",
  "threshold_adjusted",
  "session_rule_changed",
  "scoring_changed",
]);

export interface ConfigChange {
  change_category: string;
  target_timescale: string | null;
  target_tool_id: string | null;
  target_tool_type: string | null;
  target_param: string;
  old_value: unknown;
  new_value: unknown;
  reason: string;
}

function change(opts: {
  category: string;
  param: string;
  old_value?: unknown;
  new_value?: unknown;
  timescale?: string | null;
  tool_id?: string | null;
  tool_type?: string | null;
  reason?: string;
}): ConfigChange {
  return {
    change_category: opts.category,
    target_timescale: opts.timescale ?? null,
    target_tool_id: opts.tool_id ?? null,
    target_tool_type: opts.tool_type ?? null,
    target_param: opts.param,
    old_value: opts.old_value ?? null,
    new_value: opts.new_value ?? null,
    reason: opts.reason ?? "config mutation by PM agent",
  };
}

function diffScoring(
  old: Record<string, unknown>,
  new_: Record<string, unknown>,
): ConfigChange[] {
  const changes: ConfigChange[] = [];

  // thresholds
  for (const key of ["entry_threshold", "exit_threshold"]) {
    const oldVal = old[key];
    const newVal = new_[key];
    if (oldVal !== newVal) {
      changes.push(
        change({
          category: "threshold_adjusted",
          param: key,
          old_value: oldVal,
          new_value: newVal,
        }),
      );
    }
  }

  // aggregation method
  if (old.aggregation !== new_.aggregation) {
    changes.push(
      change({
        category: "scoring_changed",
        param: "aggregation",
        old_value: old.aggregation,
        new_value: new_.aggregation,
      }),
    );
  }

  // hard gate timescales
  const oldGates = (old.hard_gate_timescales as string[] | undefined) ?? [];
  const newGates = (new_.hard_gate_timescales as string[] | undefined) ?? [];
  const oldGatesSorted = [...oldGates].map(String).sort();
  const newGatesSorted = [...newGates].map(String).sort();
  if (JSON.stringify(oldGatesSorted) !== JSON.stringify(newGatesSorted)) {
    changes.push(
      change({
        category: "scoring_changed",
        param: "hard_gate_timescales",
        old_value: oldGates,
        new_value: newGates,
      }),
    );
  }

  // timescale weights
  const oldWeights =
    (old.timescale_weights as Record<string, number> | undefined) ?? {};
  const newWeights =
    (new_.timescale_weights as Record<string, number> | undefined) ?? {};
  const allTimescales = new Set([
    ...Object.keys(oldWeights),
    ...Object.keys(newWeights),
  ]);
  for (const ts of [...allTimescales].sort()) {
    const oldW = oldWeights[ts];
    const newW = newWeights[ts];
    if (oldW !== newW) {
      changes.push(
        change({
          category: "weight_adjusted",
          param: "timescale_weight",
          timescale: ts,
          old_value: oldW,
          new_value: newW,
        }),
      );
    }
  }

  return changes;
}

function diffSession(
  old: Record<string, unknown>,
  new_: Record<string, unknown>,
): ConfigChange[] {
  const changes: ConfigChange[] = [];
  const allKeys = new Set([...Object.keys(old), ...Object.keys(new_)]);
  for (const key of [...allKeys].sort()) {
    if (old[key] !== new_[key]) {
      changes.push(
        change({
          category: "session_rule_changed",
          param: key,
          old_value: old[key],
          new_value: new_[key],
        }),
      );
    }
  }
  return changes;
}

interface ToolConfig {
  instance_id: string;
  indicator_type?: string;
  action_type?: string;
  timescale?: string;
  enabled?: boolean;
  weight?: number;
  params?: Record<string, unknown>;
  [key: string]: unknown;
}

function getToolType(tool: ToolConfig, toolKind: string): string | undefined {
  return (
    (tool[`${toolKind}_type`] as string | undefined) ??
    tool.action_type ??
    tool.indicator_type
  );
}

function diffTools(
  oldTools: ToolConfig[],
  newTools: ToolConfig[],
  toolKind: string,
): ConfigChange[] {
  const changes: ConfigChange[] = [];

  const oldById = new Map<string, ToolConfig>();
  for (const t of oldTools) {
    if (t.instance_id) oldById.set(t.instance_id, t);
  }
  const newById = new Map<string, ToolConfig>();
  for (const t of newTools) {
    if (t.instance_id) newById.set(t.instance_id, t);
  }

  const allIds = new Set([...oldById.keys(), ...newById.keys()]);

  for (const toolId of [...allIds].sort()) {
    const oldTool = oldById.get(toolId);
    const newTool = newById.get(toolId);

    if (oldTool === undefined && newTool !== undefined) {
      // new tool added (enabled)
      changes.push(
        change({
          category: "tool_enabled",
          param: "enabled",
          tool_id: toolId,
          tool_type: getToolType(newTool, toolKind),
          timescale: newTool.timescale,
          old_value: null,
          new_value: true,
        }),
      );
      continue;
    }

    if (oldTool !== undefined && newTool === undefined) {
      // tool removed (disabled)
      changes.push(
        change({
          category: "tool_disabled",
          param: "enabled",
          tool_id: toolId,
          tool_type: getToolType(oldTool, toolKind),
          timescale: oldTool.timescale,
          old_value: true,
          new_value: null,
        }),
      );
      continue;
    }

    // both exist — check for changes
    if (oldTool === undefined || newTool === undefined) continue;

    const toolType = getToolType(newTool, toolKind);
    const timescale = newTool.timescale;

    // enabled/disabled toggle
    const oldEnabled = oldTool.enabled ?? true;
    const newEnabled = newTool.enabled ?? true;
    if (oldEnabled !== newEnabled) {
      const cat = newEnabled ? "tool_enabled" : "tool_disabled";
      changes.push(
        change({
          category: cat,
          param: "enabled",
          tool_id: toolId,
          tool_type: toolType,
          timescale,
          old_value: oldEnabled,
          new_value: newEnabled,
        }),
      );
    }

    // weight change (indicators only)
    if ("weight" in oldTool || "weight" in newTool) {
      const oldW = oldTool.weight;
      const newW = newTool.weight;
      if (oldW !== newW) {
        changes.push(
          change({
            category: "weight_adjusted",
            param: "weight",
            tool_id: toolId,
            tool_type: toolType,
            timescale,
            old_value: oldW,
            new_value: newW,
          }),
        );
      }
    }

    // params diff
    const oldParams = oldTool.params ?? {};
    const newParams = newTool.params ?? {};
    const allParamKeys = new Set([
      ...Object.keys(oldParams),
      ...Object.keys(newParams),
    ]);
    for (const pkey of [...allParamKeys].sort()) {
      if (oldParams[pkey] !== newParams[pkey]) {
        changes.push(
          change({
            category: "knob_tuned",
            param: pkey,
            tool_id: toolId,
            tool_type: toolType,
            timescale,
            old_value: oldParams[pkey],
            new_value: newParams[pkey],
          }),
        );
      }
    }
  }

  return changes;
}

export function computeConfigDiff(
  oldConfig: Record<string, unknown>,
  newConfig: Record<string, unknown>,
): ConfigChange[] {
  const changes: ConfigChange[] = [];

  // compare tickers
  const oldTickers = new Set(
    (oldConfig.tickers as string[] | undefined) ?? [],
  );
  const newTickers = new Set(
    (newConfig.tickers as string[] | undefined) ?? [],
  );
  if (JSON.stringify([...oldTickers].sort()) !== JSON.stringify([...newTickers].sort())) {
    changes.push(
      change({
        category: "session_rule_changed",
        param: "tickers",
        old_value: [...oldTickers].sort(),
        new_value: [...newTickers].sort(),
      }),
    );
  }

  // compare scoring
  changes.push(
    ...diffScoring(
      (oldConfig.scoring as Record<string, unknown>) ?? {},
      (newConfig.scoring as Record<string, unknown>) ?? {},
    ),
  );

  // compare session rules
  changes.push(
    ...diffSession(
      (oldConfig.session as Record<string, unknown>) ?? {},
      (newConfig.session as Record<string, unknown>) ?? {},
    ),
  );

  // compare indicators
  changes.push(
    ...diffTools(
      (oldConfig.indicators as ToolConfig[]) ?? [],
      (newConfig.indicators as ToolConfig[]) ?? [],
      "indicator",
    ),
  );

  // compare actions
  changes.push(
    ...diffTools(
      (oldConfig.actions as ToolConfig[]) ?? [],
      (newConfig.actions as ToolConfig[]) ?? [],
      "action",
    ),
  );

  return changes;
}
